libbpfgo.go

```diff
@@ -501,17 +501,23 @@ func (b *BPFMap) DeleteKey(key interface{}) error {
 	return nil
 }
 
-func (b *BPFMap) Update(key, value interface{}) error {
-	keyPtr, err := GetUnsafePointer(key)
-	if err != nil {
-		return fmt.Errorf("failed to update map %s: unknown key type %T", b.name, key)
-	}
-	valuePtr, err := GetUnsafePointer(value)
-	if err != nil {
-		return fmt.Errorf("failed to update map %s: unknown value type %T", b.name, value)
-	}
-
-	errC := C.bpf_map_update_elem(b.fd, keyPtr, valuePtr, C.BPF_ANY)
+// Update takes a pointer to a key and a value to associate it with in
+// the BPFMap. The unsafe.Pointer should be taken on a reference to the
+// underlying datatype. All basic types, and structs are supported
+//
+// NOTE: Slices and arrays are supported but references should be passed
+// to the first element in the slice or array.
+//
+// For example:
+//
+//  key := 1
+//  value := []byte{'a', 'b', 'c'}
+//  keyPtr := unsafe.Pointer(&key)
+//  valuePtr := unsafe.Pointer(&value[0])
+//  bpfmap.Update(keyPtr, valuePtr)
+//
+func (b *BPFMap) Update(key, value unsafe.Pointer) error {
```

Comment:
Should the rest of the `BPFMap` also be updated to reflect this change? E.g. `GetValue` and `DeleteKey`? That way you could delete the entire `GetUnsafePointer` function (note this will also be a breaking API change as that method is current exported).

I assume there aren't any bad implications from a consumer trying to use a pointer to a struct or an array/slice as a key, correct (or anything that specifically wasn't allowed before)?

Reply:
> Should the rest of the BPFMap also be updated to reflect this change? E.g. GetValue and DeleteKey? That way you could delete the entire GetUnsafePointer function 

+1

> I assume there aren't any bad implications from a consumer trying to use a pointer to a struct or an array/slice as a key, correct (or anything that specifically wasn't allowed before)?

With this change, the same semantics as of libbpf are applied. This means that the kernel will try to read from the given pointer the same type that was defined for the map. It is now the responsibility of the consumer to make sure that the pointer given to this function is valid and points to the correct data type, or else, bad things might happen.

Reply:
Updated GetValue and DeleteKey accordingly.